util/cache/interface.go

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package cache contains an interface for a cache around a typed value, and
+// various cache implementations that implement that interface.
+package cache
+
+import "time"
+
+// Cache is the interface for the cache types in this package.
+//
+// Functions in this interface take a key parameter, but it is valid for a
+// cache type to hold a single value associated with a key, and simply drop the
+// cached value if provided with a different key.
+//
+// It is valid for Cache implementations to be concurrency-safe or not, and
+// each implementation should document this.
+//
+// K and V should be types that can be successfully passed to json.Marshal.
```

Comment:
Although encoding/json does not (sigh), maybe we can create a `JSONMarshaler` interface (or similar name) interface, and use `V JSONMarshaler | TextMarshaler` instead of `V any`? Basically, it'd be nice if this comment were enforced at compile time.

Might also be good to report the lack of `JSONMarshaler` and the lack of its use upstream and provide a patch?

Reply:
Are you referring to something besides https://pkg.go.dev/encoding/json#Marshaler ?

It is not possible to write a static type that describes "marshalability" in general, since the marshaler walks arbitrary structures with reflection. One could eliminate certain obvious classes (e.g., functions, channels), but even then you'd have to be careful not to rule out type definitions with those underlying types.

Reply:
Yeah, my understanding is that the there's no way to express "is safe to pass to `json.Marshal`" as part of the type system, unfortunately. I wish there was!